Automattic/newspack-plugin
# feat(ras-acc): improving keyboard navigation

### All Submissions:


### Changes proposed in this Pull Request:

This PR, along with https://github.com/Automattic/newspack-blocks/pull/1760, adds some basic keyboard accessibility improvements to the reader auth and modal checkout windows. Specifically: 

* Each modal should have the attributes `role="dialog"` and `aria-modal="true"`.
* Each modal should have the modal header given an `id`, and that id should be referenced in the attribute `aria-labelledby` on the modal.
* Each modal has JavaScript that traps focus when they're opened, so when you're tabbing or shift-tabbing through the elements, the focus remains inside of the modal and doesn't go behind it.
* Each modal should be close-able with  the `esc` key.
* When each modal is closed, the focus should return to the button that opened the modal.

See 1205234045751551-as-1206049351621098.

### How to test the changes in this Pull Request:

1. Apply this PR and https://github.com/Automattic/newspack-blocks/pull/1760 and run `npm run build` for each.
2. On a page, add a donate block, a checkout button block with a single product, and a checkout button block with a product with variations. 
3. Test the following in an incognito window; do your best not to use your mouse to navigate/interact with the site. 

#### Testing Reader Registration

1. Using the 'Tab' key, navigate to the Sign In button in the header, and hit 'Enter'.
2. Confirm that the modal that opens gets focus (indicated by the bold outline on the button):

![image](https://github.com/Automattic/newspack-plugin/assets/177561/79aa916f-4f0d-41ed-9aac-778c16696292)

3. Tab through all of the elements, and confirm once you hit the bottom, you're brought back up to the close button at the top.
5. Shift-Tab backwards through the modal, and confirm once you hit the Close button, you're brought down around to the bottom.
6. Tab back to the Close button and hit 'Enter'. Confirm that the keyboard focus is returned to the Sign In button.
7. Open the Sign In button again, and proceed to sign up for a new account.
8. Once done, close the incognito window, and open a new one.
9. Use the keyboard to again navigate to the Sign In button.
10. Sign in with the same email address you just created.
11. Confirm that you get the OTP screen.
12. Repeat steps 3 and 4 to Tab and Shift tab through the modal, and confirm that focus is still trapped.
13. Confirm that the modal has the attributes `role="dialog"` and `aria-modal="true"`.
14. Confirm that the modal header has an `id`, and that id should be referenced in the attribute `aria-labelledby` on the modal.
15. Click the Esc key to close the modal, and confirm that focus returns to the Sign In button.

#### Testing Donations

1. Using the Tab key, navigate to the Donate block, and use the Tab key and arrow keys to pick a specific donation.
2. Tab to the "Donate" button, and click Enter to submit it.
3. Confirm that the modal that opens gets focus; if you're not logged in, log in.
4. On the first Billing details screen, tab through all of the elements, and confirm once you hit the bottom, you're brought back up to the close button at the top.
5. Shift-Tab backwards through the modal, and confirm once you hit the Close button, you're brought down around to the bottom.
6. Tab back to the Close button, and hit enter; confirm the focus is returned to the Donate block.
7. Reopen the Donate modal, and fill in the first screen with billing details; tab to the Continue button and hit 'Enter'.
8. On the next screen, repeat steps 4 and 5 and confirm you loop through only the visible elements.
9. With focus inside of the checkout fields (and not on the Close button), hit the 'Esc' key and confirm the focus returns to the Donate block button.
10. Reopen the Donate modal.
11. Confirm that the modal has the attributes `role="dialog"` and `aria-modal="true"`.
12. Confirm that the modal header has an `id`, and that id should be referenced in the attribute `aria-labelledby` on the modal. 
13.  Fill in the second screen. Tab to the Submit button, and click 'Enter'.
14. On the confirmation window, repeat steps 4 and 5 and confirm that focus remains in the modal.
15. Tab to the Continue button, and click 'Enter'. 
16. On the newsletter Coda screen, repeat steps 4 and 5 and confirm that focus remains in the modal.

#### Testing Checkout Button Block - simple product 

Repeat the testing steps from the Donate block, but using a Checkout Button block with a simple product.

#### Testing Checkout Button Block - variable product

Repeat the testing steps from the Donate block, but with a Checkout Button block with a variable product. The one difference is tab and shift-tab through the first variation-picker screen, and confirm that the focus remains inside of the modal.

### Other information:
